### ISSUE TITLE
Support setting kernel args for the installed OS image

### ISSUE THREAD

**Issue by abhinavdahiya** [coreos/coreos-installer#13]
When PXE booting FCOS/RHCOS the coreos-installer provides options to install OS image and the Ignition config that should be used to configure on the reboot.

There are cases where users might want to provide kernel args for the OS image that is installed by coreos-installer.

An example is:

the FCOS OS image by default has `console=ttyS0,115200n8` as the kernel parameter. But when you try to PXE boot FCOS in Packet.net `console=ttyS0,115200n8` is required to get output on the [SOS console](https://support.packet.com/kb/articles/sos-serial-over-ssh)

also an Ignition config that needs to fetch extra configuration from network, `rd.neednet=1` is required so that Ignition has DHCP/network available in initramfs.

**Comment by bgilbert** [comment]
Neither of those examples are completely on point: the kernel options for PXE-booted systems are controlled by the PXE server; the Packet platform support will include the correct console settings; and the `rd.neednet` requirement is a bug.

As much as I don't like the idea of users overriding kernel parameters, though, I think this is a good point.  @ajeddeloh, WDYT?

**Comment by ajeddeloh** [comment]
I can see wanting to set the default kernel params for an installed to disk system on BM at install time. For cloud/PXE it doesn't make much sense. That being said the implementation for diskfull BM is hairy:

Currently, with anaconda created images there's not a good way to do this without booting the system.
With BLS images (working on this) there's also no easy way to do this (could edit the BLS fragments themselves but then it'll get lost on upgrade)
Eventually we could have a grub var like `$kernel_extra_args` or something read from a snippet in `/boot` but that's a ways off.

@cgwalters was talking at some point about ostree owning the kcmdline and having it be part of the commit, but I'd argue the user should be able to augment it without needing to create new ostree commits. (I agree the _defaults_ ought to be part of the commit).

**Comment by dustymabe** [comment]
> I can see wanting to set the default kernel params for an installed to disk system on BM at install time.

yep. I asked @abhinavdahiya to open this issue after chatting with him. I thought it was a reasonable feature request.

**Comment by bgilbert** [comment]
One use case for kargs: configuring SMT for mitigating L1TF and MDS.  That does apply more to bare metal than VMs, but bare-metal clouds that provide preinstalled images (e.g. Packet) will need it.  Thus, doing it from the installer isn't really enough; it should be possible to set kargs from Ignition.

**Comment by cgwalters** [comment]
I actually argued for kargs in Ignition but Andrew talked me out of it.

https://github.com/openshift/machine-config-operator/issues/388 is the MCO issue here, and has an open PR now.  For RHCOS we are likely to just rely on the fact that we generally always need to do an initial pivot (i.e. upgrade and reboot) and use that as the opportunity to change kernel args.

I guess an interesting question here is whether Ignition should also support "upgrade and reboot" - I would be shocked if this hadn't come up in the CL context.  If Ignition supports that then it also argues for kargs?

**Comment by arithx** [comment]
I'd currently lean towards having a service in the initramfs with a conditional requirement that if met performs a reboot. With that in place we could implement a way of writing kargs via Ignition without adding specific logic to Ignition itself and use sugar in the distro implementation of CT to simplify the process for users.

**Comment by cgwalters** [comment]
One thing to remember if we pursue this is to move `ignition-firstboot-complete.service` from the real rootfs into Ignition itself so we don't run Ignition twice.

**Comment by ajeddeloh** [comment]
My current thinking is write a flag file to like `/tmp/needs-reboot` or something similar to trigger it. Would be a behavior of the initramfs not Ignition.

**Comment by bgilbert** [comment]
@cgwalters I'm not aware of Ignition upgrade+reboot ever being discussed for CL.  https://github.com/coreos/fedora-coreos-tracker/issues/34 is relevant though.

@ajeddeloh I was hoping that the unit would detect the actual file written by Ignition that requires the reboot (`/etc/kcmdline` or something).  Requiring the config to write out a separate flag file seems awkward.

**Comment by bgilbert** [comment]
Filed https://github.com/coreos/ignition-dracut/issues/81 for rebooting from the initramfs.

**Comment by cgwalters** [comment]
I realized in https://github.com/openshift/machine-config-operator/pull/762/ that it'd really be a lot better if we handed kargs in a way integrated with Ignition, since otherwise we need to carry "MCD parts" in RHCOS that runs very early.

**Comment by cgwalters** [comment]
> @cgwalters I'm not aware of Ignition upgrade+reboot ever being discussed for CL. coreos/fedora-coreos-tracker#34 is relevant though.

For RHCOS to share this mechanism fully we also need upgrades.  Now...I am just hesitating a lot on pushing the whole stack into the initramfs versus the alternative of booting to a special `coreos-firstboot.target` or so, and running whatever we need there from the real root, before we start other services.  This would mean that any 3rd parties dropping units that are `ConditionFirstBoot=` would have to explicitly hook them in there.

**Comment by bgilbert** [comment]
> Now...I am just hesitating a lot on pushing the whole stack into the initramfs versus the alternative of booting to a special coreos-firstboot.target or so, and running whatever we need there from the real root, before we start other services.

@cgwalters Could you expand on that?  (Maybe in https://github.com/coreos/fedora-coreos-tracker/issues/34.)  To me the initramfs seems strictly better, for the same reason that Ignition runs there.

**Comment by cgwalters** [comment]
With the new "unified" metal/BIOS no-grub metal disk images in FCOS (but not RHCOS), the canonical source of the kernel arguments is in `/boot/loader`.  That makes having coreos-installer change them rather easy.

**Comment by dustymabe** [comment]
agree.. some related discussion: https://github.com/coreos/fedora-coreos-tracker/issues/191#issuecomment-519346594

**Comment by jlebon** [comment]
So... now that we have the new coreos-installer, one could just mount `/boot` post-install (and pre-reboot) and change the BLS configs. Though... the presence of `--firstboot-args` is kinda asking for a similar `--persistent-args` facilitator.

**Comment by jlebon** [comment]
I guess this gets back to https://github.com/coreos/ignition-dracut/issues/81 though; if we have kargs handling built-in, then we should just use that so it's the same workflow across cloud and bare metal.

**Comment by jlebon** [comment]
There's some overlap with the auto-forward magic we do with the network kargs too though. For example, if `console=` kargs were passed in for the install boot, it's very likely that users want those kargs to stick around too. (The difference though with the network kargs is that we want it on every boot, not just on the first boot). Tempting to auto-forward that too, though need to balance the UX with keeping magic at a minimum.

**Comment by brokenjacobs** [comment]
Just to add to this, we have the exact same issue installing on bare metal. IPMI on our supermicro bare metal platform requires console=ttyS1,115200n8 not console=ttyS0,115200n8. And using the append option at install time does not work on the first reboot. So we are in the dark until we are able to use rpm-ostree kargs --replace... in a systemd unit on the first reboot. Not ideal.

Creating an additional 'platform' for a very common bare metal use case seems like a lot of work just to have control over basic kernel arguments.
